docs/plugin-protocol/tfplugin6.1.proto

```diff
@@ -0,0 +1,332 @@
+            string attribute_name = 1;
+            // Set "element_key_*" to represent looking up an element in
+            // an indexable collection type.
+            string element_key_string = 2;
+            int64 element_key_int = 3;
+        }
+    }
+    repeated Step steps = 1;
+}
+
+message StopProvider {
+    message Request {
+    }
+    message Response {
+        string Error = 1;
+    }
+}
+
+// RawState holds the stored state for a resource to be upgraded by the
+// provider. It can be in one of two formats, the current json encoded format
+// in bytes, or the legacy flatmap format as a map of strings.
+message RawState {
+    bytes json = 1;
+    map<string, string> flatmap = 2;
+}
+
+enum StringKind {
+    PLAIN = 0;
+    MARKDOWN = 1;
+}
+
+// Schema is the configuration schema for a Resource or Provider.
+message Schema {
+    message Block {
+        int64 version = 1;
+        repeated Attribute attributes = 2;
+        repeated NestedBlock block_types = 3;
+        string description = 4;
+        StringKind description_kind = 5;
+        bool deprecated = 6;
+    }
+
+    message Attribute {
+        string name = 1;
+        bytes type = 2;
+        Object nested_type = 10;
+        string description = 3;
+        bool required = 4;
+        bool optional = 5;
+        bool computed = 6;
+        bool sensitive = 7;
+        StringKind description_kind = 8;
+        bool deprecated = 9;
+    }
+
+    message NestedBlock {
+        enum NestingMode {
+            INVALID = 0;
+            SINGLE = 1;
+            LIST = 2;
+            SET = 3;
+            MAP = 4;
+            GROUP = 5;
+        }
+
+        string type_name = 1;
+        Block block = 2;
+        NestingMode nesting = 3;
+        int64 min_items = 4;
+        int64 max_items = 5;
+    }
+
+    message Object {
+        enum NestingMode {
+            INVALID = 0;
+            SINGLE = 1;
+            LIST = 2;
+            SET = 3;
+            MAP = 4;
+        }
+
+        repeated Attribute attributes = 1;
+        NestingMode nesting = 3;
+        int64 min_items = 4;
+        int64 max_items = 5;
+    }
+
+    // The version of the schema.
+    // Schemas are versioned, so that providers can upgrade a saved resource
+    // state when the schema is changed. 
+    int64 version = 1;
+
+    // Block is the top level configuration block for this schema.
+    Block block = 2;
+
+    // Capability allows providers to communicate extra information regarding
+    // the behavior of individual resources. This is used to indicate
+    // forward-compatible changes which cannot be directly tested by the plugin
+    // client, allowing protocol5 resources to be muxed with strict protocol6
+    // resources from the same provider.
+    //
+    // Current valid values are
+    // 0: Legacy protocol5 compatible resource
+    // 1: Protocol6 resource which will not use SchemaConfigModeAttr
+    int64 capability = 3;
```

Comment:
I like the idea of capability negotiation, but it feels weird to pack it all into just a single 64-bit integer like this, cause that puts constraints on how we could use it in future.

What do you think about defining a `message Capability` which for now just has a single field `bool structural_attrs` which, when enabled, turns on the ability to use structural attribute types in return for _disabling_ the legacy attributes as blocks mode?

(I'm implicitly suggesting here that it would be invalid to use `Attribute.nested_type` _without_ activating this capability, so that these two behaviors are forced to be mutually exclusive and thus we can avoid dealing with the inevitable quirky interactions between these two modes -- but I feel more strongly about using a message type rather than and integer than I do about what exactly the flag inside does.)

Then if we have new capabilities in future we can add them as separate fields with their own docs. Some of them might have types other than `bool`, depending on what we end up needing to represent.

Reply:
I guess we need to decide if we actually want ever add more capabilities based via the schema alone. I guess I brought this on by picking the work `Capability` over some form of "MetaVersion" :D  If we do actually want "capabilities" stored in here in the long run, then I see no reason to not add this is an a new message type. 

I don't think we can force this to be opt-in for the _new_ behavior however, since there could technically be providers built using the existing mechanisms which would be broken by disabling nested types, and the default zero value should not exclude those. Though if you think we can get away with that somehow, I am all for making the behavior as strict as possible.

Reply:
Since this is only prevent undesirable but not necessarily broken behavior, I see no need to backport this to 1.0, which gives the SDKs time to implement and get releases in place to opt-in. It does feel kind of strange that we implemented a new protocol to add the additional features, only to require an additional flag to opt-in to said features, but it is really to exclude old behavior to it's not _that_ awful.

If that solution is ok with @paddycarver, I'll add `Schema.Capabilities.NestedTypes` to the protocol and re-work the PR accordingly.

Reply:
My sense was that the protocol version change handled the other end of this negotiation, preventing odd behavior if someone tries to use a newer provider with an older version of Terraform. In retrospect it seems like we could've potentially added a more fine-grain mechanism for _that_ too, but the new protocol version gave us an opportunity to clean up some other oddities we learned about since 0.12 anyway so I think it was overall worth the effort.

Reply:
And here I missed the less vague term "StructuralAttrs" :/, the PR cleanup will also `s/NestedTypes/StructuralAttrs/`